### PR TITLE
feat: Asana migrator

### PR DESCRIPTION
**Pull request by Anurag810**
TODO:
- [x] Development.
- [ ] Test cases. (Not required as we can not expose API Key)

This feature allows you to import all your projects from each and every workspace to your Bloomsatck instance on click.

<h3>Asana Settings</h3>
Store your Personal Acess Token from Aasana required to authorize your request to fetch data from your Asana Instance.

Screenshot from 2021-10-05 12-19-15

<h3>Asana Migrator</h3>
This Document type helps you to select and edit your task details during import. 
This is one line introduction (If required).

**Edit Task during import:**

Editable-Grid

**Real-time progress while fetching Tasks:**

Peek 2021-10-05 12-23-1

<h3> Tasks </h3>
It fetches all details like description, Assignee, Reference and, Comments.

Screenshot from 2021-10-05 14-11-47

It also set up parent-child relations for tasks accordingly
<h3>Workflow:</h3>

- Set your Personal Acess Token in Asana Settings and save 
- Go to Asana migrator
    - Click on New, It will automatically fetch workspaces and projects for you.
Peek 2021-10-05 12-21
    - Click on save.
- Click on get Task.
- it will fetch and render all tasks from the project in the data table.
- Edit and select tasks to import from the data table.
- click on Import task.
- The project and Task will be created. 

Workflow

### CODE REVIEW

**Comment by Anurag810** [comment]
@neilLasrado I don't know, what to put on the test because there is no prospect for a developer/sandbox environment that developers can use in tests and exposing API Key to test files is not good practice.

**Comment by neilLasrado** [comment]
@sahil28297 can you pull this on bloomstack.garden to test ?

**Comment by neilLasrado** [comment]
We need to fix the name for this. @DeveloperMetal @rvpasari @sahil28297 @vishdha any suggestions?

I was thinking 'Asana Data Import'... In the future tools that we build to import data from other software can be named similarly eg- METRC Data Import, QuickBooks Data Import etc

File: erpnext/erpnext_integrations/doctype/asana_migrator/asana_migrator.js [hunk 1]
```diff
@@ -0,0 +1,304 @@
...
+					args: {
+						tasks_to_import,
+						comments_to_import
+					},
+					callback: function () {
+						frm.refresh();
+					}
+				});
+			} else {
+				frappe.msgprint(__("Please select Tasks from Datatable to import"));
```
Comment (neilLasrado) [inline erpnext/erpnext_integrations/doctype/asana_migrator/asana_migrator.js]: Please select Tasks from Table to import

Comment (neilLasrado) [reply erpnext/erpnext_integrations/doctype/asana_migrator/asana_migrator.js]: Use user-friendly msg prints

File: erpnext/erpnext_integrations/doctype/asana_migrator/asana_migrator.js [hunk 2]
```diff
@@ -0,0 +1,304 @@
...
+				frm.doc.workspace_options = JSON.stringify(r.message);
+			}
+		});
+	},
+
+	get_projects: function (frm, global_id) {
+		frappe.call({
+			method: "erpnext.erpnext_integrations.doctype.asana_migrator.asana_migrator.get_projects",
+			freeze: true,
+			freeze_message: __("Getting Projects..."),
```
Comment (neilLasrado) [inline erpnext/erpnext_integrations/doctype/asana_migrator/asana_migrator.js]: Fetching Projects from Asana...

File: erpnext/erpnext_integrations/doctype/asana_migrator/asana_migrator.js [hunk 3]
```diff
@@ -0,0 +1,304 @@
...
+
+		// saving Options as Workspace value as options are Dynamic.
+		frm.doc.selected_workspace = frm.doc.workspace;
+	},
+
+	get_workspaces: function (frm) {
+		frappe.call({
+			method: "erpnext.erpnext_integrations.doctype.asana_migrator.asana_migrator.get_workspaces",
+			freeze: true,
+			freeze_message: __("Getting Workspaces..."),
```
Comment (neilLasrado) [inline erpnext/erpnext_integrations/doctype/asana_migrator/asana_migrator.js]: Fetching Workspaces from Asana...

File: erpnext/erpnext_integrations/doctype/asana_migrator/asana_migrator.js [hunk 4]
```diff
@@ -0,0 +1,304 @@
...
+	make_datatable: function (frm) {
+		let editable = true;
+		if (frm.doc.status == "Completed") {
+			editable = false;
+		}
+		let columns = frm.events.get_columns(editable);
+		let row = JSON.parse(frm.doc.row_data);
+		frm.get_field("tasks_view").$wrapper.append(`<div>
+			<span class="indicator whitespace-nowrap blue">
+				<span>Sub-Tasks will be imported Automatically</span>
```
Comment (neilLasrado) [inline erpnext/erpnext_integrations/doctype/asana_migrator/asana_migrator.js]: Select Tasks that you wish to import from the table below. Sub-Tasks will be imported Automatically

File: erpnext/erpnext_integrations/doctype/asana_migrator/asana_migrator.js (hunk 1 above)
Comment (Anurag810) [reply erpnext/erpnext_integrations/doctype/asana_migrator/asana_migrator.js]: Done

File: erpnext/erpnext_integrations/doctype/asana_migrator/asana_migrator.js (hunk 2 above)
Comment (Anurag810) [reply erpnext/erpnext_integrations/doctype/asana_migrator/asana_migrator.js]: Done

File: erpnext/erpnext_integrations/doctype/asana_migrator/asana_migrator.js (hunk 3 above)
Comment (Anurag810) [reply erpnext/erpnext_integrations/doctype/asana_migrator/asana_migrator.js]: Done

**Comment by Anurag810** [comment]
@neilLasrado we cannot do anything about this. we require DataTable initialization, sider is putting that out.
Screenshot from 2021-10-20 11-49-05

**Review by neilLasrado — APPROVED** [review]

**Pull request merged**